community/kernel/src/main/java/org/neo4j/kernel/impl/store/id/FreeIdKeeper.java

```diff
@@ -22,90 +22,122 @@
+ * primarily because the aggressiveMode argument influences exactly that behaviour.
+ * <p>
+ * The {@link #aggressiveMode} parameter controls whether or not IDs which are freed during this lifecycle will
+ * be allowed to be reused during the same lifecycle. The alternative non-aggressive behaviour is that the IDs
+ * will only be reused after a close/open cycle. This would generally correlate with a restart of the database.
  */
 public class FreeIdKeeper implements Closeable
 {
-    public static final long NO_RESULT = -1;
-    public static final int ID_ENTRY_SIZE = Long.BYTES;
-    private final LinkedList<Long> freeIds = new LinkedList<>();
-    private final LinkedList<Long> readFromDisk = new LinkedList<>();
+    private static final int ID_ENTRY_SIZE = Long.BYTES;
+
+    private final List<Long> freeIds = new ArrayList<>();
+    private final List<Long> readFromDisk = new ArrayList<>();
     private final StoreChannel channel;
-    private final int threshold;
-    /*
-     * aggressiveReuse flags if ids freed during this run (before close() is called) are legitimate return values or not.
-     * If yes, then they are shown preference over ids persisted to disk. If not, then they are persisted in batches
-     * but are not returned until this keeper is closed and reopened. This is achieved by marking the position of the file
-     * where the persisted ids run up to when the file was opened and never reading past that. The newly freed ids then
-     * are persisted beyond that point and are never read.
+    private final int batchSize;
+    private final boolean aggressiveMode;
+
+    private long freeIdCount;
+
+    /**
+     * Keeps the position where batches of IDs will be flushed out to.
+     * This can be viewed as being put on top of a stack.
      */
-    private final boolean aggressiveReuse;
-    private long defraggedIdCount;
+    private long stackPosition;
 
-    private final long lowWatermarkForChannelPosition; // the lowest possible position the channel can be at - we don't own anything "in front" of that
-    /*
-     * maxReadPosition remains constant if aggressiveReuse is false, pointing to the position after which we should not read because
-     * it contains overflow ids from this run. If aggressiveReuse is true, then it points to the end of the file.
+    /**
+     * The position before we started this run.
+     *
+     * Useful to keep track of the gap that will form in non-aggressive mode
+     * when IDs from old runs get reused and newly freed IDs are put on top
+     * of the stack. During a clean shutdown the gap will be compacted away.
+     *
+     * During an aggressive run a gap is never formed since batches of free
+     * IDs are flushed on top of the stack (end of file) and also read in
+     * from the top of the stack.
      */
-    private long maxReadPosition;
-    private long readPosition; // the place from where we read. Always <= maxReadPosition
+    private long prePosition;
```

Comment:
I would call it something like `initialPosition` instead

Reply:
same same